File: solver/search_bench.py
"""Compare native search builds with exact root-stat parity and paired timings.

  python solver/search_bench.py /path/before.so solver/fast_engine.so

Uses identical opening/midgame positions, private dice, fixed/adaptive budgets,
play/grading modes, and alternating execution order. No strength tradeoff is
accepted: actions, visits, Qs, simulation totals, and the game RNG must match.
"""
import argparse
import importlib.util
import os
from pathlib import Path
import statistics
import time

import numpy as np


def load(path, name):
    previous = os.environ.get('NW_ENGINE_SO')
    os.environ['NW_ENGINE_SO'] = str(Path(path).resolve())
    try:
        spec = importlib.util.spec_from_file_location(name, Path(__file__).with_name('fastnw.py'))
        module = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(module)
        return module
    finally:
        if previous is None:
            os.environ.pop('NW_ENGINE_SO', None)
        else:
            os.environ['NW_ENGINE_SO'] = previous


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument('before')
    ap.add_argument('after')
    ap.add_argument('--seeds', type=int, default=12)
    ap.add_argument('--sims', type=int, default=4000)
    ap.add_argument('--repeats', type=int, default=3)
    args = ap.parse_args()
    engines = [load(args.before, 'before'), load(args.after, 'after')]
    elapsed = [[], []]
    positions = 0
    for seed in range(1, args.seeds + 1):
        base = engines[0]
        g = base.new_game(seed)
        # End-turn transitions yield deterministic early and midgame positions,
        # including difficult/wiped positions; keep only live decisions.
        for turn in range(1, 5):
            if base.check_winner(g['owner']) >= 0 or not np.any(g['owner'] == 0):
                break
            positions += 1
            for grade in (0, 1):
                for repeat in range(args.repeats):
                    results = [None, None]
                    for k in ((0, 1) if repeat % 2 == 0 else (1, 0)):
                        e = engines[k]
                        e.set_topology_csr(g['n'], g['adj'])
                        e.use_mb32(g['mb'])
                        e.use_sim(0x12345678)
                        e.set_grade(grade)
                        start = time.perf_counter()
                        results[k] = e.uct_search(g['owner'], g['strength'], turn,
                            args.sims, return_q=True,
                            max_sims=args.sims * (2 if repeat == 1 else 1))
                        elapsed[k].append(time.perf_counter() - start)
                        assert e.get_mb32() == g['mb'], 'search changed game dice'
                    for a, b in zip(results[0], results[1]):
                        assert np.array_equal(a, b), (seed, turn, grade, repeat)
                    assert engines[0].sims_done() == engines[1].sims_done()
            base.use_mb32(g['mb'])
            base.end_turn(g['owner'], g['strength'])
            g['mb'] = base.get_mb32()
    total = list(map(sum, elapsed))
    ratio = statistics.median(a / b for a, b in zip(*elapsed))
    print(f'PASS: {positions} positions, {len(elapsed[0])} paired searches; exact root stats and RNG')
    print(f'Before {total[0]:.3f}s; after {total[1]:.3f}s; '
          f'aggregate speedup {total[0]/total[1]:.3f}x; median paired speedup {ratio:.3f}x')


if __name__ == '__main__':
    main()

File: solver/search_bench.mjs
// Compare two WASM builds with paired timings and exact root-stat/RNG parity.
// node solver/search_bench.mjs /path/before.mjs public/fast_engine.js [seeds=12] [sims=4000]
import assert from 'node:assert/strict';
import { readFile } from 'node:fs/promises';
import { pathToFileURL } from 'node:url';
import { resolve } from 'node:path';
const [before, after, seedArg = '12', simsArg = '4000'] = process.argv.slice(2);
if (!before || !after) throw new Error('Provide before and after WASM module paths');
const wrapper = await readFile(new URL('../public/fastnw.js', import.meta.url), 'utf8');
const engines = [];
for (const path of [before, after]) {
  const source = wrapper.replace("'./fast_engine.js'", JSON.stringify(pathToFileURL(resolve(path)).href));
  const { loadEngine } = await import('data:text/javascript;base64,' + Buffer.from(source).toString('base64'));
  engines.push(await loadEngine());
}
const elapsed = [0, 0], ratios = [], sims = Number(simsArg);
let positions = 0, pairs = 0;
for (let seed = 1; seed <= Number(seedArg); seed++) {
  const base = engines[0], g = base.newGame(seed);
  for (let turn = 1; turn <= 4; turn++) {
    if (base.checkWinner(g.owner) >= 0 || !g.owner.includes(0)) break;
    positions++;
    for (const grade of [0, 1]) for (let repeat = 0; repeat < 3; repeat++) {
      const results = [], times = [];
      for (const k of (repeat % 2 ? [1, 0] : [0, 1])) {
        const e = engines[k];
        e.setTopologyCsr(g.n, g.adj); e.useMb32(g.mb);
        e.useSim(0x12345678); e.setGrade(grade);
        const start = performance.now();
        results[k] = e.uctSearch(g.owner, g.strength, turn, sims, 2.5, 1,
          sims * (repeat === 1 ? 2 : 1));
        times[k] = performance.now() - start; elapsed[k] += times[k];
        assert.equal(e.getMb32(), g.mb);
      }
      assert.deepEqual(results[1], results[0], `seed ${seed}, turn ${turn}, grade ${grade}`);
      assert.equal(engines[0].uctSimsDone(), engines[1].uctSimsDone());
      ratios.push(times[0] / times[1]); pairs++;
    }
    base.useMb32(g.mb);
    base._put(base._owner, g.owner); base._put(base._strength, g.strength);
    base.M._end_turn(base._owner, base._strength);
    base._getBack(base._owner, g.owner); base._getBack(base._strength, g.strength);
    g.mb = base.getMb32();
  }
}
ratios.sort((a,b) => a-b);
console.log(`PASS: ${positions} positions, ${pairs} paired searches; exact root stats and RNG`);
console.log(`Before ${(elapsed[0]/1000).toFixed(3)}s; after ${(elapsed[1]/1000).toFixed(3)}s; aggregate speedup ${(elapsed[0]/elapsed[1]).toFixed(3)}x; median paired speedup ${ratios[ratios.length>>1].toFixed(3)}x`);
// Reserve the actual browser ceiling to measure allocated linear memory as well.
for (const e of engines) {
  const g = e.newGame(1); e.uctBegin(g.owner, g.strength, 1, 2000, 2.5, 1, 150000);
}
console.log('WASM heap at 150k ceiling (MiB), before/after:', engines.map(e => (e.M.HEAP32.buffer.byteLength / 2**20).toFixed(2)).join(' / '));
